LUCENE-9983: Stop sorting determinize powersets unnecessarily

zhaih opened the pull request:
<!--
_(If you are a project committer then you may remove some/all of the following template.)_

Before creating a pull request, please file an issue in the ASF Jira system for Lucene:

* https://issues.apache.org/jira/projects/LUCENE

You will need to create an account in Jira in order to create an issue.

The title of the PR should reference the Jira issue number in the form:

* LUCENE-####: <short description of problem or changes>

LUCENE must be fully capitalized. A short description helps people scanning pull requests for items they can work on.

Properly referencing the issue in the title ensures that Jira is correctly updated with code review comments and commits. -->


# Description

In `determinize` function (which truns NFA to DFA), we use `SortedIntSet` to store powerset and it uses a sorted data structure which could be very costly in some cases. And we actually don't need to keep the set always sorted.

# Solution

Replace usage of `SortedIntSet` by `StateSet`, a wrapper of `IntIntHashMap`. Note I haven't removed `SortedIntSet` but probably we should since it is only used in `determinize`?

# Tests
 * Reused old tests that used to test `SortedIntSet`
 * I also applied the patch attached in [LUCENE-9981](https://issues.apache.org/jira/projects/LUCENE/issues/LUCENE-9981). Before the commit it runs 6min before throw an exception and after this commit it is 16 sec.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/lucene/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Lucene maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.

zhaih committed:
LUCENE-9983: Stop sorting determinize powersets unnecessarily

Replace usage of SortedIntSet by StateSet, a wrapper of IntIntHashMap

mikemccand commented on lucene/core/build.gradle:
```diff
@@ -20,6 +20,8 @@ apply plugin: 'java-library'
 description = 'Lucene core library'
 
 dependencies {
+  implementation 'com.carrotsearch:hppc'
```
Can you add a `// nocommit` here?  We can't take (external) dependencies in Lucene's `core` ... but hopefully we can fork the one HPPC class we need here (`IntIntHashMap`?).

zhaih replied:
Oh I see, we might need to modify the source so that we won't bring in a whole interface set used in HPPC.

dweiss replied:
Feel free to cut and trim to your will. This is exactly why it's licensed the way it is. @bruno-roustant came up with some clever new hashing improvements recently - these are not published as a public revision but you can get them from the repository and compile it locally. See this for details:

https://issues.carrot2.org/browse/HPPC-176

bruno-roustant replied:
According to the stats you shared @zhaih in the Jira issue, there are 3 times more get (inc/dec) & removal than entry additions. If there are less than 2M states then IntIntWormMap should be more efficient.
I suggest that you benchmark with IntIntHashMap and IntIntWormMap (with your existing tests) and take the most appropriate.

zhaih replied:
@bruno-roustant I tried WormMap yesterday (with hppc 0.9.0.RC2) and I didn't see benefits from the adversarial test case. Just to educate myself, is removal also a fast operation?

bruno-roustant replied:
Yes removal is fast.

mikemccand commented on lucene/core/src/java/org/apache/lucene/util/automaton/StateSet.java:
```diff
@@ -0,0 +1,107 @@
+final class StateSet extends IntSet {
+
+  private final IntIntHashMap inner;
+  private int hashCode;
```
@dweiss suggested upgrading the hash to `long` to further reduce chance of false collisions.

zhaih replied:
Ah yeah that's a good idea, I forgot to integrate it here, will do.

dweiss replied:
Not just moving to a long but also using a better hash function. The typical accumulative default  (prime * current + elementValue) is fine but if the number of hashed elements is small (or if their values are small) this leads to poor distributions. Throw a murmur mix function in between (or at the end at least) and things typically look much better.

mikemccand commented on lucene/core/src/java/org/apache/lucene/util/automaton/StateSet.java:
```diff
@@ -0,0 +1,107 @@
+    changed = false;
+    arrayCache = inner.keys().toArray();
+    // we need to sort this array since "equals" method depend on this
+    Arrays.sort(arrayCache);
```
Oh I see, we are still sorting on `freeze`, but not when accumulating the powerset.  Curious this is so much faster :)  I would expect it to be slower on the adversarial regexp, I think, because we pay the sort price on every freeze now, instead on insert/delete into this map?  Confusing :)

zhaih replied:
I have posted some stats in the JIRA issue, it turns out we are doing way way much more add/delete than freeze. Since  `(#add + #delete)/#freeze > 10000` so a rough analysis is each `add` or `delete` use `O(log(n))` and `freeze` use `O(n * log(n))`, we see the cost of `#add + #delete` is still larger, though I'm not sure why it is this much faster.

zhaih replied:
Oh another thought is, we were using raw array for states less than 30, which have `add/delete` cost of `O(n)`, and I would expect most of the powerset will be constructed using these 30 states (since `C(30,4) > 10000` already), meaning we might be actually paying `O(n)` price per `add/delete` most times

dweiss replied:
Yeah... I think a good empirical breakdown of what actually is happening may be much better than theoretical complexity analysis. This doesn't mean it's not worth experimenting with int-int hash maps but I have a gut feeling you can get pretty darn close with less memory used using existing sorted arrays.

bruno-roustant replied:
How many states are we manipulating? If it's not too many, then indeed maybe a simple array of state-counters is enough?
If array[i] is the counter for state i, then incr(i) / decr(i) are 0(1), hash is still O(n) (skipping 0 values), freeze does not need to be sorted.

mikemccand replied:
> How many states are we manipulating? If it's not too many, then indeed maybe a simple array of state-counters is enough?
> If array[i] is the counter for state i, then incr(i) / decr(i) are 0(1), hash is still O(n) (skipping 0 values), freeze does not need to be sorted.

That's a neat idea!

The tricky thing is that these are usually very sparsely occupied sets (each DFA state is "typically" just a few NFA states, I think), so we would have to pay the O(N) cost to walk that array and turn it into the frozen (sparse `int[]`) representation.

mikemccand reviewed (commented):
Whoa, thank you for the quick initial PR @zhaih!

It looks like we are still sorting, during `freeze` -- it's curious this is so much faster in your test!

dweiss commented on lucene/core/src/java/org/apache/lucene/util/automaton/StateSet.java:
```diff
@@ -0,0 +1,107 @@
+    }
+    hashCode = inner.size();
+    for (IntCursor cursor : inner.keys()) {
+      hashCode += BitMixer.mix(cursor.value);
```
+1. An alternative: 37 * hashCode + cursor.value with a final mix step.

bruno-roustant commented on lucene/core/src/java/org/apache/lucene/util/automaton/Operations.java:
```diff
@@ -676,7 +677,7 @@ public static Automaton determinize(Automaton a, int maxDeterminizedStates) {
 
     // Same initial values and state will always have the same hashCode
-    FrozenIntSet initialset = new FrozenIntSet(new int[] {0}, 683, 0);
+    FrozenIntSet initialset = new FrozenIntSet(new int[] {0}, BitMixer.mix(0) + 1, 0);
```
Why not simply 0 as initial hash code?

zhaih replied:
Just to keep the hash code the same with the one used in `StateSet`, 0 should be the hash code used for 0 length array I think?

bruno-roustant commented on lucene/core/src/java/org/apache/lucene/util/automaton/StateSet.java:
```diff
@@ -0,0 +1,107 @@
+  // Adds this state to the set
+  void incr(int num) {
+    if (inner.addTo(num, 1) == 1) {
+      changed = true;
```
Could we rename "changed" to "keysChanged" as this only tracks keys and not values?

zhaih replied:
+1

bruno-roustant commented on lucene/core/src/java/org/apache/lucene/util/automaton/StateSet.java:
```diff
@@ -0,0 +1,107 @@
+    int keyIndex = inner.indexOf(num);
+    int count = inner.indexGet(keyIndex) - 1;
+    if (count == 0) {
+      inner.remove(num);
```
@dweiss ideally here I would use inner.indexRemove(keyIndex). Does this miss in HPPC?

zhaih replied:
Thanks Bruno for adding this capability! I'll fork the indexRemove and use it here.

bruno-roustant commented on lucene/core/src/java/org/apache/lucene/util/automaton/StateSet.java:
```diff
@@ -0,0 +1,107 @@
+    }
+  }
+
+  void computeHash() {
```
I find strange that this method is called externally by Operations. Since the hash is cached and we track when it needs to be recomputed with "changed", could we make it private and change the signature to "private int getOrComputeHash()"?

mikemccand replied:
Yeah this is odd -- `+1` to make the hash code a privately maintained thing.

mikemccand commented on lucene/core/src/java/org/apache/lucene/util/automaton/StateSet.java:
```diff
@@ -0,0 +1,107 @@
+   */
+  FrozenIntSet freeze(int state) {
+    if (changed == false) {
+      assert arrayCache != null;
```
Hmm do we actually fall inside this `if`?  I would think we shouldn't ever hit this -- we should't call `freeze` unless something had in fact changed?  Or, if we are, something else might be wrong?  Or maybe I am simply confused ;)

dweiss replied:
bq. why hasn't this been "published in a public revision" yet :)

Don't know. Life gets in the way.

zhaih replied:
@mikemccand We actually might fall inside this `if`. So before we call `freeze`, we will perform a `get` using this `StateSet` by the `newStates` hashmap, there we might call `getArray()` if there's hashcode collision and `changed` should be set to `false` there.

mikemccand reviewed (commented):
I suspect most `FrozenIntSet`s in "good" cases are very small, and we might see gains by leaving unsorted when they are less than N (3?  4?) in size, and doing the inefficient O(N^2) comparison on hash collisions/true equals.

But we badly need a "realistic" benchmark of regexps as a compass to navigate all these fun optimizations we are talking about here!

zhaih commented:
Thank you @mikemccand @dweiss and @bruno-roustant all for reviewing this PR! Since this PR is more of an optimization for adversarial cases so we don't want to sacrifice performance of our normal use cases. I will take some time to benchmark those changes (this one as well as few others that are not yet presented) first and see what numbers we'll get to decide how to move on with this PR.

zhaih committed:
Make hashcode for IntSet classes long
make StateSet's computeHashCode private
Forked HPPC IntIntHashMap and BitMixer, as well as tests for IntIntHashMap

dweiss commented on lucene/core/src/java/org/apache/lucene/util/automaton/StateSet.java:
```diff
@@ -0,0 +1,106 @@
+    }
+    hashCode = inner.size();
+    for (IntIntHashMap.IntCursor cursor : inner.keys()) {
+      hashCode += BitMixer.mix(cursor.value);
```
I'd still think the simple aggregation 37 * hashCode + cursor.value with a final mix step would be a good enough (or maybe even better) choice here.

zhaih replied:
But wouldn't that make hashcode depend on order of keys?

dweiss replied:
Duh. Of course! Sorry for not noticing this.

dweiss commented on lucene/core/src/java/org/apache/lucene/util/BitMixer.java:
```diff
@@ -0,0 +1,130 @@
+ * Bit mixing utilities. The purpose of these methods is to evenly distribute key space over int32
+ * range.
+ *
+ * <p>Forked from com.carrotsearch.hppc.BitMixer
```
I'd make it a link back to the original project and include a comment which hash ref of the repository this was copied from, for reference and in case somebody wishes to sync/update in the future.

zhaih replied:
Done!

mikemccand reviewed (commented):
Net/net this looks great!

We stop sorting while incrementally accumulating/maintaining the powerset (incr/decr), by switching to HPPC int/int map, and only sort on `freeze`, which is only called per determinized (output) state.  For costly NFAs, which have large powersets, with way more incr/decr than eventual DFA state count, such as the adversarial regex that led to this, this is a massive speedup!

And for "normal" regexes, with not so many NFA states in each powerset, this should be performance neutral.

So I think this is a win/win!  I left some small comments but I think this is close!  Thanks @zhaih!

zhaih committed:
Move HPPC related classes to o.a.l.hppc package
Also forked indexRemove method

bruno-roustant commented on lucene/core/src/java/org/apache/lucene/util/automaton/Operations.java:
```diff
@@ -693,7 +694,7 @@ public static Automaton determinize(Automaton a, int maxDeterminizedStates) {
     final PointTransitionSet points = new PointTransitionSet();
 
     // like SortedMap<Integer,Integer>
```
The comment should be replaced.

bruno-roustant commented on lucene/core/src/java/org/apache/lucene/util/automaton/StateSet.java:
```diff
@@ -0,0 +1,106 @@
+import org.apache.lucene.util.hppc.BitMixer;
+import org.apache.lucene.util.hppc.IntIntHashMap;
+
+/** A thin wrapper of {@link IntIntHashMap} */
```
Should we say something about the keys and values?
key = state num, value = ref count

bruno-roustant commented on lucene/core/src/java/org/apache/lucene/util/automaton/StateSet.java:
```diff
@@ -0,0 +1,106 @@
+    inner = new IntIntHashMap(capacity);
+  }
+
+  // Adds this state to the set
```
Should be javadoc. Increments the ref count of a state?

bruno-roustant commented on lucene/core/src/java/org/apache/lucene/util/automaton/StateSet.java:
```diff
@@ -0,0 +1,106 @@
+    }
+  }
+
+  // Removes this state from the set, if count decrs to 0
```
Better with javadoc.

bruno-roustant commented on lucene/core/src/test/org/apache/lucene/util/automaton/TestIntSet.java:
```diff
@@ -31,14 +31,13 @@ public void testFreezeEqualityLargeSet() {
 
   private void testFreezeEquality(int size) {
-    SortedIntSet sortedSet = new SortedIntSet(0);
+    StateSet sortedSet = new StateSet(0);
```
Can we rename sortedSet? (confusing)

bruno-roustant reviewed (approved):
Nice! I just have some small comments left about doc.

zhaih committed:
Javadoc improvement

zhaih committed:
Move the PR to 8.10.0

mikemccand reviewed (approved):
Thanks @zhaih this looks great -- I think it's ready!  Have you confirmed that `gradle  check` passes?

zhaih commented:
> Thanks @zhaih this looks great -- I think it's ready! Have you confirmed that `gradle check` passes?

Thank you for reviewing it! Yeah `gradle check` passed.

mikemccand commented:
> > Thanks @zhaih this looks great -- I think it's ready! Have you confirmed that `gradle check` passes?
> 
> Thank you for reviewing it! Yeah `gradle check` passed.

Great!  I'll try to push this today!  Thanks @zhaih!

mikemccand committed:
Merge branch 'main' into main
